### PR TITLE
SOLR-16506: Throws exception on core startup for node name mismatch

### PR DESCRIPTION
**Pull request by patsonluk**
https://issues.apache.org/jira/browse/SOLR-16506

# Description

Detailed description in the Jira link provided above.

TL;DR:
During core registration on node startup, discrepancy in node name defined in state.json from ZK could result as accidental data deletion on the node that hosts the actual replica

# Solution

It's risky to proceed with publishing replica as down and updating state.json if the node name is not a match. We are proposing throwing a new exception `InconsistentClusterStateException` in `ZkController#checkStateInZk` to interrupt the core registration process if there's hostname mismatch.

# Remarks
In a way the old code could have "auto-corrected" the state.json in some scenarios (for example a jetty instance is restarted with a different port number in unit test case, or node with ephemeral port is expected to restart and locate the replicas created with different port number?) 

Therefore this PR will remove such "auto-correct" behavior. I do believe that such behavior can actually be dangerous as described, therefore the proposed new exception.

Would love to hear opinions from other devs in the community! 😊 

# Tests

Several existing test cases are updated to reuse the port number during restart. Otherwise, the new exception would cause failures to those unit test cases.

Added a new test `TestCoreInconsistentReplica` to confirm that:
- it should NOT publish to state.json to "autocorrect" the node name
- it should NOT interfere with other core creations that have correct node name

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 1]
```diff
@@ -2899,6 +2920,23 @@ public NotInClusterStateException(ErrorCode code, String msg) {
     }
   }
 
+  /**
+   * Thrown during pre-register process if the replica is found in clusterstate but is inconsistent
+   * with current node name
+   */
+  public static class InconsistentClusterStateException extends SolrException {
```
Comment (noblepaul) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:2927]: Why a new Exception? The `replica` field is not used anywhere

**Review by noblepaul — CHANGES_REQUESTED** [review]

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 1 above)
Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:2927]: The new exception better identifies a new type of exception. This cannot be `NotInClusterStateException`, otherwise the `CoreContainer` will attempt to unload (ie delete)  the core at https://github.com/fullstorydev/lucene-solr/blob/release/8.8/solr/core/src/java/org/apache/solr/core/CoreContainer.java#L1525.

The extra field is just to store the info which could be useful for any future handling/logging

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:2927]: My concern is that this Exception is not caught anywhere or we do not do an `instanceof` check

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:2927]: Since this is a non runtime exception, the caller should be forced to handle such exception specifically or this should fall to any handling of unexpected exceptions of existing logic? :)

In this case, the caller `preRegister` does already have handling for general unexpected exception, so this is supposed to work as is - be wrapped at `SolrException`, re-thrown and caught in `createFromDescriptor`, but unlike `NotInClusterStateException`, which unload(delete) the physical directory, it will just interrupt the core loading, wrapped and thrown again, and finally be logged in `CoreContainer#load` without affecting other cores.

Such described flow is already the handling of any unexpected exception which I find quite fitting for our scenario. And there's an added test case to ensure a single `InconsistentClusterStateException ` should not affect loading of other "healthy" cores. 

Lemme know if you have any thoughts on different handling!

**Comment by HoustonPutman** [comment]
How does this interact with the `genericCoreNodeNames` option in [the solr.xml](https://solr.apache.org/guide/solr/latest/configuration-guide/configuring-solr-xml.html#the-solrcloud-element)? It seems like that is made so that cores can be moved between nodes...

**Comment by patsonluk** [comment]
> How does this interact with the `genericCoreNodeNames` option in [the solr.xml](https://solr.apache.org/guide/solr/latest/configuration-guide/configuring-solr-xml.html#the-solrcloud-element)? It seems like that is made so that cores can be moved between nodes...

Thank you for the raising the concern @HoustonPutman , I was not aware of such option! From my brief search the reference on `ZkController#genericCoreNodeNames`/`cloudConfig.getGenericCoreNodeNames()`, it only appears to affect the result of `getCoreNodeName`. So I assume the `Replica#getNodeName` (read from the state.json, which `genericCoreNodeNames` should not matter?) would still be a mismatch to the `ZkController#getNodeName`.

However, ur concern does lead me to think deeper on several scenarios, which I think would be great if you can provide some insights:
1. From `genericCoreNodeNames`, it mentions the case of `When a different machine takes over serving that core things will be much easier to understand.`. I assume all replica modes should be done via admin APIs (ie MOVESHARD/REPLACENODE) etc, which should update the state.json accordingly. Do we support some other forms of moves? like physically moving the folders around and expect Solr to detect the change and "self correct" the state.json? My approach assumes that if state.json is incorrect, we should not attempt to self-correct it, but I could have misunderstood some of the designs 😓 
2. What about node name changes on a host? For example if we used to pass IP as `-Dhost` but now with the hostname, should we assume that Solr would start up fine and state.json will be "self-corrected"?

**Comment by patsonluk** [comment]
Mismatching node name could happen when `-Dhost` is used, under such circumstances, it is desirable to publish a down state with the updated node name (this probably explains the [check](https://github.com/apache/solr/blob/11253f05cfb31f9fb945c831d8889b3db1e607f1/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L2035) here).

Closing this proposal as this change could prevent the correct handling.
